File: project/test_segmentation.py
"""
Test script for Irish text segmentation functionality
Run with: python test_segmentation.py
"""

import re
import time

class IrishEReader:
    def __init__(self):
        self.current_sentence_index = 0
        self.sentences = []
        self.is_playing = False
        self.is_practice_mode = False

    def clean_input_text(self, text):
        # Normalize whitespace
        text = re.sub(r'\s+', ' ', text)
        # Remove excessive line breaks
        text = re.sub(r'\n{3,}', '\n\n', text)
        # Fix spacing around punctuation
        text = re.sub(r'\s*([.!?])\s*', r'\1 ', text)
        # Remove leading/trailing whitespace
        return text.strip()

    def detect_sentence_boundaries(self, text):
        boundaries = [0]
        
        irish_abbreviations = {
            'srl', 'rl', 'ucht', 'm.sh', 'dr', 'teo', 'teoranta',
            'lch', 'lch.', 'caib', 'caib.', 'b.á.c', 'bác',
            'eag', 'eag.', 'cpt', 'cpt.', 'lth', 'lth.'
        }

        common_abbreviations = {
            'mr', 'mrs', 'ms', 'dr', 'prof', 'st', 'ave', 'rd',
            'inc', 'ltd', 'co', 'corp', 'etc', 'vs', 'ie', 'eg',
            'am', 'pm', 'a.m', 'p.m'
        }

        all_abbreviations = irish_abbreviations.union(common_abbreviations)
        
        # Find sentence-ending punctuation
        for match in re.finditer(r'[.!?]+', text):
            position = match.end()
            
            if self.is_valid_sentence_boundary(text, position, all_abbreviations):
                boundaries.append(position)

        if boundaries[-1] != len(text):
            boundaries.append(len(text))

        return boundaries

    def is_valid_sentence_boundary(self, text, position, abbreviations):
        before_context = text[max(0, position - 10):position]
        after_context = text[position:min(len(text), position + 10)]

        # Check for abbreviations
        if self.is_abbreviation(before_context, abbreviations):
            return False

        # Check for decimal numbers
        if re.search(r'\d\.\d', before_context + (after_context[:1] if after_context else '')):
            return False

        # Check for URLs or email addresses
        if re.search(r'[a-zA-Z]\.[a-zA-Z]', before_context + after_context):
            return False

        # Must have whitespace or end of text after punctuation
        next_char = text[position] if position < len(text) else ''
        if next_char and not re.match(r'\s', next_char):
            return False

        # If next character exists, it should be uppercase or a quote
        next_non_whitespace = after_context.strip()[:1] if after_context else ''
        if (next_non_whitespace and 
            not re.match(r'[A-ZÁÉÍÓÚÀÈÌÒÙÂÊÎÔÛÄËÏÖÜŶĆÑŇ"""'']', next_non_whitespace)):
            return False

        return True

    def is_abbreviation(self, context, abbreviations):
        words = context.lower().split()
        if not words:
            return False
            
        last_word = words[-1]
        clean_word = re.sub(r'[.!?]+$', '', last_word)
        return clean_word in abbreviations

    def count_words(self, text):
        return len([word for word in text.strip().split() if word])

    def detect_irish_content(self, text):
        irish_characters = re.search(r'[áéíóúàèìòùâêîôûäëïöüŷńñ]', text, re.IGNORECASE)
        irish_words = re.search(
            r'\b(agus|le|ar|i|in|an|na|is|tá|bhí|go|do|de|sa|sna|den|don|faoi|ó|trí|chun|mar|leis|léi|dá|má|ach|nó|mura|sula|nuair|cén|cad|conas|cá|cathain|cé|céard)\b',
            text, re.IGNORECASE
        )
        
        return bool(irish_characters or irish_words)

    def extract_punctuation(self, text):
        return re.findall(r'[.!?;:,"""''()[\]{}—–-]', text)

    def create_sentence_objects(self, text, boundaries):
        sentences = []

        for i in range(len(boundaries) - 1):
            start = boundaries[i]
            end = boundaries[i + 1]
            
            content = text[start:end].strip()
            
            if content:
                sentence = {
                    'id': i,
                    'content': content,
                    'start_position': start,
                    'end_position': end,
                    'length': len(content),
                    'word_count': self.count_words(content),
                    'has_irish_content': self.detect_irish_content(content),
                    'punctuation': self.extract_punctuation(content),
                    'is_valid': True
                }
                sentences.append(sentence)

        return sentences

    def filter_valid_sentences(self, sentences):
        valid_sentences = []
        
        for i, sentence in enumerate(sentences):
            # Remove empty or too short sentences
            if not sentence['content'] or len(sentence['content']) < 2:
                continue

            # Remove sentences that are just punctuation or numbers
            if re.match(r'^[.!?;:,\s\d]+$', sentence['content']):
                continue

            # Handle very long sentences
            if sentence['word_count'] > 200:
                sentence['is_long'] = True
                print(f"Warning: Long sentence detected: {sentence['word_count']} words")

            sentence['number'] = len(valid_sentences) + 1
            sentence['id'] = len(valid_sentences)
            valid_sentences.append(sentence)

        return valid_sentences

    def parse_text_to_sentences(self, text):
        if not text or not isinstance(text, str):
            return []

        cleaned_text = self.clean_input_text(text)
        sentence_boundaries = self.detect_sentence_boundaries(cleaned_text)
        sentences = self.create_sentence_objects(cleaned_text, sentence_boundaries)
        valid_sentences = self.filter_valid_sentences(sentences)
        
        print(f"Parsed {len(valid_sentences)} valid sentences from text")
        return valid_sentences


def run_tests():
    e_reader = IrishEReader()
    tests_run = 0
    tests_passed = 0

    print("=== Irish Text Segmentation Tests ===\n")

    # Test 1: Basic Irish text
    print("Test 1: Basic Irish text")
    test1_text = "Tá mé ag foghlaim na Gaeilge. Is breá liom an teanga seo! Cad é mar atá tú?"
    test1_result = e_reader.parse_text_to_sentences(test1_text)
    tests_run += 1
    
    if len(test1_result) == 3:
        print("✅ PASS: Correctly identified 3 sentences")
        tests_passed += 1
    else:
        print(f"❌ FAIL: Expected 3 sentences, got {len(test1_result)}")

    irish_sentences = [s for s in test1_result if s['has_irish_content']]
    tests_run += 1
    if len(irish_sentences) == 3:
        print("✅ PASS: All sentences detected as Irish content")
        tests_passed += 1
    else:
        print(f"❌ FAIL: Expected 3 Irish sentences, got {len(irish_sentences)}")

    print()

    # Test 2: Abbreviations
    print("Test 2: Irish abbreviations")
    test2_text = "Bhí mé ag caint le Dr. Ó Síocháin inné. Fuair mé litir ó Ucht. na hÉireann."
    test2_result = e_reader.parse_text_to_sentences(test2_text)
    tests_run += 1
    
    if len(test2_result) == 2:
        print("✅ PASS: Correctly handled abbreviations")
        tests_passed += 1
    else:
        print(f"❌ FAIL: Expected 2 sentences, got {len(test2_result)}")
        for i, s in enumerate(test2_result):
            print(f"  {i + 1}: \"{s['content']}\"")

    print()

    # Test 3: Edge cases
    print("Test 3: Edge cases")
    test3_text = "Numbers like 3.14 are decimals. URLs like www.google.com should not split. Email test@example.com too."
    test3_result = e_reader.parse_text_to_sentences(test3_text)
    tests_run += 1
    
    if len(test3_result) == 3:
        print("✅ PASS: Correctly handled edge cases")
        tests_passed += 1
    else:
        print(f"❌ FAIL: Expected 3 sentences, got {len(test3_result)}")
        for i, s in enumerate(test3_result):
            print(f"  {i + 1}: \"{s['content']}\"")

    print()

    # Test 4: Mixed language
    print("Test 4: Mixed language")
    test4_text = "Hello, my name is Seán. Táim ag foghlaim na Gaeilge. I live in Dublin."
    test4_result = e_reader.parse_text_to_sentences(test4_text)
    tests_run += 1
    
    if len(test4_result) == 3:
        print("✅ PASS: Correctly parsed mixed language")
        tests_passed += 1
    else:
        print(f"❌ FAIL: Expected 3 sentences, got {len(test4_result)}")

    mixed_irish = [s for s in test4_result if s['has_irish_content']]
    tests_run += 1
    if len(mixed_irish) == 1:
        print("✅ PASS: Correctly identified 1 Irish sentence in mixed text")
        tests_passed += 1
    else:
        print(f"❌ FAIL: Expected 1 Irish sentence, got {len(mixed_irish)}")

    print()

    # Test 5: Performance test
    print("Test 5: Performance test")
    large_text = "Tá mé ag foghlaim na Gaeilge. " * 1000
    start_time = time.time()
    performance_result = e_reader.parse_text_to_sentences(large_text)
    end_time = time.time()
    processing_time = (end_time - start_time) * 1000  # Convert to milliseconds

    tests_run += 1
    if len(performance_result) == 1000 and processing_time < 1000:
        print(f"✅ PASS: Performance test completed in {processing_time:.2f}ms")
        tests_passed += 1
    else:
        print(f"❌ FAIL: Performance issue - {processing_time:.2f}ms for {len(performance_result)} sentences")

    print()

    # Summary
    print("=== Test Summary ===")
    print(f"Tests run: {tests_run}")
    print(f"Tests passed: {tests_passed}")
    print(f"Tests failed: {tests_run - tests_passed}")
    print(f"Success rate: {(tests_passed / tests_run * 100):.1f}%")

    if tests_passed == tests_run:
        print("\n🎉 All tests passed! Irish text segmentation is working correctly.")
        return True
    else:
        print("\n❌ Some tests failed. Check the implementation.")
        return False


if __name__ == "__main__":
    success = run_tests()
    
    # Demonstrate usage
    if success:
        print("\n=== Usage Example ===")
        reader = IrishEReader()
        sample_text = """
        Fáilte go dtí an léitheoir Gaeilge! Tá sé seo cruthaithe chun cabhrú leat do chuid fuaimnithe a fheabhsú. 
        Is féidir leat téacs a ionchuir anseo, agus déanfaidh an córas anailís air. 
        Beidh tú in ann éisteacht leis an téacs agus do ghutha féin a thaifeadadh.
        """
        
        sentences = reader.parse_text_to_sentences(sample_text)
        print(f"\nParsed {len(sentences)} sentences:")
        for sentence in sentences:
            print(f"  {sentence['number']}: {sentence['content'][:60]}{'...' if len(sentence['content']) > 60 else ''}")
            print(f"    Words: {sentence['word_count']}, Irish: {sentence['has_irish_content']}")

File: project/test-segmentation.js
/**
 * Test script for Irish text segmentation functionality
 * Run with: node test-segmentation.js
 */

// Mock DOM elements for testing
global.document = {
    getElementById: () => null,
    createElement: () => ({ 
        className: '', 
        textContent: '', 
        style: {}, 
        addEventListener: () => {},
        appendChild: () => {}
    }),
    addEventListener: () => {}
};

global.window = {};
global.localStorage = {
    getItem: () => null,
    setItem: () => {}
};

global.alert = console.log;
global.console = console;

// Import the e-reader class (we'll need to modify it slightly for testing)
class IrishEReader {
    constructor() {
        this.currentSentenceIndex = 0;
        this.sentences = [];
        this.isPlaying = false;
        this.isPracticeMode = false;
    }

    cleanInputText(text) {
        return text
            .replace(/\s+/g, ' ')
            .replace(/\n{3,}/g, '\n\n')
            .replace(/\s*([.!?])\s*/g, '$1 ')
            .trim();
    }

    detectSentenceBoundaries(text) {
        const boundaries = [0];
        
        const irishAbbreviations = new Set([
            'srl', 'rl', 'ucht', 'm.sh', 'dr', 'teo', 'teoranta',
            'lch', 'lch.', 'caib', 'caib.', 'b.á.c', 'bác',
            'eag', 'eag.', 'cpt', 'cpt.', 'lth', 'lth.'
        ]);

        const commonAbbreviations = new Set([
            'mr', 'mrs', 'ms', 'dr', 'prof', 'st', 'ave', 'rd',
            'inc', 'ltd', 'co', 'corp', 'etc', 'vs', 'ie', 'eg',
            'am', 'pm', 'a.m', 'p.m'
        ]);

        const allAbbreviations = new Set([...irishAbbreviations, ...commonAbbreviations]);
        const sentenceEndPattern = /[.!?]+/g;
        let match;

        while ((match = sentenceEndPattern.exec(text)) !== null) {
            const position = match.index + match[0].length;
            
            if (this.isValidSentenceBoundary(text, position, allAbbreviations)) {
                boundaries.push(position);
            }
        }

        if (boundaries[boundaries.length - 1] !== text.length) {
            boundaries.push(text.length);
        }

        return boundaries;
    }

    isValidSentenceBoundary(text, position, abbreviations) {
        const beforeContext = text.substring(Math.max(0, position - 10), position);
        const afterContext = text.substring(position, Math.min(text.length, position + 10));

        if (this.isAbbreviation(beforeContext, abbreviations)) {
            return false;
        }

        if (/\d\.\d/.test(beforeContext + afterContext.charAt(0))) {
            return false;
        }

        if (/[a-zA-Z]\.[a-zA-Z]/.test(beforeContext + afterContext)) {
            return false;
        }

        const nextChar = text.charAt(position);
        if (nextChar && !/\s/.test(nextChar)) {
            return false;
        }

        const nextNonWhitespace = afterContext.trim().charAt(0);
        if (nextNonWhitespace && 
            !/[A-ZÁÉÍÓÚÀÈÌÒÙÂÊÎÔÛÄËÏÖÜŶĆÑŇ"""'']/u.test(nextNonWhitespace)) {
            return false;
        }

        return true;
    }

    isAbbreviation(context, abbreviations) {
        const words = context.toLowerCase().split(/\s+/);
        const lastWord = words[words.length - 1];
        
        if (!lastWord) return false;

        const cleanWord = lastWord.replace(/[.!?]+$/, '');
        return abbreviations.has(cleanWord);
    }

    countWords(text) {
        return text.trim().split(/\s+/).filter(word => word.length > 0).length;
    }

    detectIrishContent(text) {
        const irishCharacters = /[áéíóúàèìòùâêîôûäëïöüŷńñ]/i;
        const irishWords = /\b(agus|le|ar|i|in|an|na|is|tá|bhí|go|do|de|sa|sna|den|don|faoi|ó|trí|chun|mar|leis|léi|dá|má|ach|nó|mura|sula|nuair|cén|cad|conas|cá|cathain|cé|céard)\b/i;
        
        return irishCharacters.test(text) || irishWords.test(text);
    }

    extractPunctuation(text) {
        const punctuation = text.match(/[.!?;:,"""''()[\]{}—–-]/g) || [];
        return punctuation;
    }

    createSentenceObjects(text, boundaries) {
        const sentences = [];

        for (let i = 0; i < boundaries.length - 1; i++) {
            const start = boundaries[i];
            const end = boundaries[i + 1];
            
            const content = text.substring(start, end).trim();
            
            if (content) {
                const sentence = {
                    id: i,
                    content: content,
                    startPosition: start,
                    endPosition: end,
                    length: content.length,
                    wordCount: this.countWords(content),
                    hasIrishContent: this.detectIrishContent(content),
                    punctuation: this.extractPunctuation(content),
                    isValid: true
                };

                sentences.push(sentence);
            }
        }

        return sentences;
    }

    filterValidSentences(sentences) {
        return sentences.filter(sentence => {
            if (!sentence.content || sentence.content.length < 2) {
                return false;
            }

            if (/^[.!?;:,\s\d]+$/.test(sentence.content)) {
                return false;
            }

            if (sentence.wordCount > 200) {
                sentence.isLong = true;
                console.warn(`Long sentence detected: ${sentence.wordCount} words`);
            }

            sentence.number = sentences.indexOf(sentence) + 1;
            return true;
        }).map((sentence, index) => {
            sentence.number = index + 1;
            sentence.id = index;
            return sentence;
        });
    }

    parseTextToSentences(text) {
        if (!text || typeof text !== 'string') {
            return [];
        }

        const cleanedText = this.cleanInputText(text);
        const sentenceBoundaries = this.detectSentenceBoundaries(cleanedText);
        const sentences = this.createSentenceObjects(cleanedText, sentenceBoundaries);
        const validSentences = this.filterValidSentences(sentences);
        
        console.log(`Parsed ${validSentences.length} valid sentences from text`);
        return validSentences;
    }
}

// Test cases
function runTests() {
    const eReader = new IrishEReader();
    let testsRun = 0;
    let testsPassed = 0;

    console.log('=== Irish Text Segmentation Tests ===\n');

    // Test 1: Basic Irish text
    console.log('Test 1: Basic Irish text');
    const test1Text = "Tá mé ag foghlaim na Gaeilge. Is breá liom an teanga seo! Cad é mar atá tú?";
    const test1Result = eReader.parseTextToSentences(test1Text);
    testsRun++;
    
    if (test1Result.length === 3) {
        console.log('✅ PASS: Correctly identified 3 sentences');
        testsPassed++;
    } else {
        console.log(`❌ FAIL: Expected 3 sentences, got ${test1Result.length}`);
    }

    const irishSentences = test1Result.filter(s => s.hasIrishContent);
    if (irishSentences.length === 3) {
        console.log('✅ PASS: All sentences detected as Irish content');
        testsPassed++;
    } else {
        console.log(`❌ FAIL: Expected 3 Irish sentences, got ${irishSentences.length}`);
    }
    testsRun++;

    console.log('\n');

    // Test 2: Abbreviations
    console.log('Test 2: Irish abbreviations');
    const test2Text = "Bhí mé ag caint le Dr. Ó Síocháin inné. Fuair mé litir ó Ucht. na hÉireann.";
    const test2Result = eReader.parseTextToSentences(test2Text);
    testsRun++;
    
    if (test2Result.length === 2) {
        console.log('✅ PASS: Correctly handled abbreviations');
        testsPassed++;
    } else {
        console.log(`❌ FAIL: Expected 2 sentences, got ${test2Result.length}`);
        test2Result.forEach((s, i) => console.log(`  ${i + 1}: "${s.content}"`));
    }

    console.log('\n');

    // Test 3: Edge cases
    console.log('Test 3: Edge cases');
    const test3Text = "Numbers like 3.14 are decimals. URLs like www.google.com should not split. Email test@example.com too.";
    const test3Result = eReader.parseTextToSentences(test3Text);
    testsRun++;
    
    if (test3Result.length === 3) {
        console.log('✅ PASS: Correctly handled edge cases');
        testsPassed++;
    } else {
        console.log(`❌ FAIL: Expected 3 sentences, got ${test3Result.length}`);
        test3Result.forEach((s, i) => console.log(`  ${i + 1}: "${s.content}"`));
    }

    console.log('\n');

    // Test 4: Mixed language
    console.log('Test 4: Mixed language');
    const test4Text = "Hello, my name is Seán. Táim ag foghlaim na Gaeilge. I live in Dublin.";
    const test4Result = eReader.parseTextToSentences(test4Text);
    testsRun++;
    
    if (test4Result.length === 3) {
        console.log('✅ PASS: Correctly parsed mixed language');
        testsPassed++;
    } else {
        console.log(`❌ FAIL: Expected 3 sentences, got ${test4Result.length}`);
    }

    const mixedIrish = test4Result.filter(s => s.hasIrishContent);
    testsRun++;
    if (mixedIrish.length === 1) {
        console.log('✅ PASS: Correctly identified 1 Irish sentence in mixed text');
        testsPassed++;
    } else {
        console.log(`❌ FAIL: Expected 1 Irish sentence, got ${mixedIrish.length}`);
    }

    console.log('\n');

    // Test 5: Performance test
    console.log('Test 5: Performance test');
    const largeText = "Tá mé ag foghlaim na Gaeilge. ".repeat(1000);
    const startTime = process.hrtime.bigint();
    const performanceResult = eReader.parseTextToSentences(largeText);
    const endTime = process.hrtime.bigint();
    const processingTime = Number(endTime - startTime) / 1000000; // Convert to milliseconds

    testsRun++;
    if (performanceResult.length === 1000 && processingTime < 1000) {
        console.log(`✅ PASS: Performance test completed in ${processingTime.toFixed(2)}ms`);
        testsPassed++;
    } else {
        console.log(`❌ FAIL: Performance issue - ${processingTime.toFixed(2)}ms for ${performanceResult.length} sentences`);
    }

    console.log('\n');

    // Summary
    console.log('=== Test Summary ===');
    console.log(`Tests run: ${testsRun}`);
    console.log(`Tests passed: ${testsPassed}`);
    console.log(`Tests failed: ${testsRun - testsPassed}`);
    console.log(`Success rate: ${(testsPassed / testsRun * 100).toFixed(1)}%`);

    if (testsPassed === testsRun) {
        console.log('\n🎉 All tests passed! Irish text segmentation is working correctly.');
    } else {
        console.log('\n❌ Some tests failed. Check the implementation.');
    }
}

// Run the tests
runTests();